### PR TITLE
[HIVE-21686] ensure that memory allocator does not evict using brute …

### PR DESCRIPTION
**Pull request by b-slim**
[HIVE-21686] ensure that memory allocator does not evict using brute foce path.

Change-Id: I020d9be9e76ac095f909b53cbec9e63f21efe8ff

### CODE REVIEW

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java [hunk 1]
```diff
@@ -277,7 +301,74 @@ public void allocateMultiple(MemoryBuffer[] dest, int size, BufferObjectFactory
...
+      if (forceEvictAttempt >= MAX_FAST_ATTEMPT) {
+        // Try to evict more starting from how much memory we are missing allocLog2 << dest.length - destAllocIx
+        // Exponentially increase the eviction request based on forceEvictAttempt counter.
+        // Why Exponentially increase, usually query fragments triggers a burst of allocation at the same time
+        // Thus IMO to avoid lock contentions it is worth to over evict thus next allocation will get free lunch.
+        currentEvicted =
+            memoryManager.evictMemory(allocLog2 << dest.length - destAllocIx + forceEvictAttempt - MAX_FAST_ATTEMPT);
+        // Note that memoryManager.evictMemory is blocking but eventually will return 0 if all the memory is locked
+        // thus mark empty attempts and bail out after MAX_DISCARD_ATTEMPTS
+        totalForceEvictedBytes += currentEvicted;
```
Comment (odraese) [inline llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java]: This while loop might also exit with a single (large) forced eviction? Specifically if the number of unallocated buffers cause a large shift of allocLog2, the totalForceEvictedBytes might get > maxForcedEvictionSize and therefore leave the loop without the chance of using any of the evicted memory.

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java [hunk 2]
```diff
@@ -294,25 +385,18 @@ public void allocateMultiple(MemoryBuffer[] dest, int size, BufferObjectFactory
...
-    int attempt = 0;
+    int discardsAttempt = 0;
     boolean isFailed = false;
     int memoryForceReleased = 0;
     try {
       int discardFailed = 0;
       while (true) {
         // Try to split bigger blocks.
-        int startArenaIx = (int)((threadId + attempt) % arenaCount);
+        startArenaIx = (int)((threadId + discardsAttempt) % arenaCount);
```
Comment (odraese) [inline llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java]: Maybe we want to add forceEvictAttempt here, not to start at the first (threadID given) arena again?

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java [hunk 3]
```diff
@@ -352,18 +440,21 @@ public void allocateMultiple(MemoryBuffer[] dest, int size, BufferObjectFactory
...
           String msg = "Failed to allocate " + size + "; at " + destAllocIx + " out of "
               + dest.length + " (entire cache is fragmented and locked, or an internal issue)";
           logOomErrorMessage(msg);
           throw new AllocatorOutOfMemoryException(msg);
         }
-        ++attempt;
+        ++discardsAttempt;
       }
     } finally {
+      //@TODO CHECK i think it is a bug, we do not have to release this memory if we already did deallocate(dest[i])
```
Comment (odraese) [inline llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java]: The deallocate path doesn't increment memoryForceReleased, so this here should be fine.

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/LowLevelCacheMemoryManager.java [hunk 4]
```diff
@@ -66,6 +66,21 @@ public void reserveMemory(final long memoryToReserve, AtomicBoolean isStopped) {
     throw new ReserveFailedException(isStopped);
   }
 
+  /**
+   * Ask the memory manager to evict more memory
+   *
+   * @param memoryToReserve amount of bytes to evict
+   * @return actual amount of evicted bytes.
+   */
+  @Override public long evictMemory(long memoryToReserve) {
```
Comment (odraese) [inline llap-server/src/java/org/apache/hadoop/hive/llap/cache/LowLevelCacheMemoryManager.java]: memoryToReserve -> memoryToEvict?

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/MemoryManager.java [hunk 5]
```diff
@@ -57,4 +57,13 @@
...
    */
   void reserveMemory(long memoryToReserve, AtomicBoolean isStopped);
+
+  /**
+   * Ask the memory manager to evict more memory
+   *
+   * @param memoryToReserve amount of bytes to evict
+   * @return actual amount of evicted bytes.
+   */
+  long evictMemory(long memoryToReserve);
```
Comment (odraese) [inline llap-server/src/java/org/apache/hadoop/hive/llap/cache/MemoryManager.java]: Adding this to MemoryManager kind'o breaks the separation of memory manager and eviction policy (at least from interface perspective, these were two separate things so far).

File: llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestBuddyAllocator.java [hunk 6]
```diff
@@ -87,7 +91,7 @@ public void testSameSizes() throws Exception {
     int min = 3, max = 8, maxAlloc = 1 << max;
     BuddyAllocator a = new BuddyAllocator(isDirect, isMapped, 1 << min, maxAlloc, maxAlloc,
         maxAlloc, 0, tmpDir, new DummyMemoryManager(),
-        LlapDaemonCacheMetrics.create("test", "1"), null, true);
+        LlapDaemonCacheMetrics.create("test", "1"), null, true, 1024 * 1024 * 128);
```
Comment (odraese) [inline llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestBuddyAllocator.java]: You're using this 128MB value multiple times in this suite. Maybe, you want to move this into a const.

File: llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestCacheAllocationsEvictionsCycles.java [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    cachePolicy.purge();
+    allocator = null;
+    memoryManager = null;
+    cachePolicy = null;
+  }
+
+  /**
+   * Test case to ensure that deallocate it does merge small blocks into bigger ones.
+   */
+  @Test(timeout = 6_000L) public void TestMergeOfBlocksAfterDeallocate() {
```
Comment (odraese) [inline llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestCacheAllocationsEvictionsCycles.java]: Seems, that all other tests (that I read so far) in LLAP followed the standard Java method name convention, to start with a lowercase t.

File: llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestOrcMetadataCache.java [hunk 8]
```diff
@@ -102,7 +104,7 @@ public void testBuffers() throws Exception {
     final int MAX_ALLOC = 64;
     LlapDaemonCacheMetrics metrics = LlapDaemonCacheMetrics.create("", "");
     BuddyAllocator alloc = new BuddyAllocator(
-        false, false, 8, MAX_ALLOC, 1, 4096, 0, null, mm, metrics, null, true);
+        false, false, 8, MAX_ALLOC, 1, 4096, 0, null, mm, metrics, null, true, 1024 * 1024 * 128);
```
Comment (odraese) [inline llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestOrcMetadataCache.java]: Same here. 128M could be a final static...

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java [hunk 9]
```diff
@@ -277,7 +301,74 @@ public void allocateMultiple(MemoryBuffer[] dest, int size, BufferObjectFactory
...
+        return;
+      }
+
+      if (forceEvictAttempt >= MAX_FAST_ATTEMPT) {
+        // Try to evict more starting from how much memory we are missing allocLog2 << dest.length - destAllocIx
+        // Exponentially increase the eviction request based on forceEvictAttempt counter.
+        // Why Exponentially increase, usually query fragments triggers a burst of allocation at the same time
+        // Thus IMO to avoid lock contentions it is worth to over evict thus next allocation will get free lunch.
+        currentEvicted =
+            memoryManager.evictMemory(allocLog2 << dest.length - destAllocIx + forceEvictAttempt - MAX_FAST_ATTEMPT);
```
[the remark this answers is not in the document]
Comment (b-slim) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java]: good catch @odraese

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 10]
```diff
@@ -4053,6 +4053,9 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
     LLAP_ALLOCATOR_DEFRAG_HEADROOM("hive.llap.io.allocator.defrag.headroom", "1Mb",
         "How much of a headroom to leave to allow allocator more flexibility to defragment.\n" +
         "The allocator would further cap it to a fraction of total memory."),
+    LLAP_ALLOCATOR_MAX_FORCE_EVICTED("hive.llap.io.allocator.max.force.eviction", "16Mb",
+        "How much bytes we can force evict before using free List discards and/or burt force\n" +
```
[the remark this answers is not in the document]
Comment (b-slim) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: did a reword let me know if you think it is still unclear.

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java (hunk 1 above)
Comment (b-slim) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java]: we have more tries down at allocator discard  section, but i agree can be moved up.

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/MemoryManager.java (hunk 5 above)
Comment (b-slim) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/MemoryManager.java]: Memory manager is doing eviction on reserve call, thus this is adding an explicit handle to the eviction without reserving it. do you a different idea?

File: llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestCacheAllocationsEvictionsCycles.java (hunk 7 above)
Comment (b-slim) [reply llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestCacheAllocationsEvictionsCycles.java]: fixed.

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java (hunk 2 above)
Comment (b-slim) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java]: done

**Comment by b-slim** [comment]
@odraese Thanks for the review, i did address most of you comment can you please take a second look when you have time.

**Comment by b-slim** [comment]
thanks pushed to master

**Pull request closed without merging**